cluster/encryption.go

```diff
@@ -113,94 +116,202 @@ func (c *Cluster) DisableSecretsEncryption(ctx context.Context, currentCluster *
+	rewritten := make(map[string]struct{})
+	var rmtx sync.RWMutex
+
+	// spawn workers to perform secret rewrites
+	var errgrp errgroup.Group
 	for w := 0; w < SyncWorkers; w++ {
 		errgrp.Go(func() error {
 			var errList []error
-			for secret := range secretsQueue {
+			for secret := range rewrites {
 				s := secret.(v1.Secret)
-				err := rewriteSecret(k8sClient, &s)
-				if err != nil {
-					errList = append(errList, err)
+				id := getSecretID(s)
+
+				rmtx.RLock()
+				_, ok := rewritten[id]
+				rmtx.RUnlock()
+
+				if !ok {
+					err := rewriteSecret(k8sClient, &s)
+					if err != nil {
+						errList = append(errList, err)
+					}
+
+					rmtx.Lock()
+					rewritten[id] = struct{}{}
+					rmtx.Unlock()
+
+					done <- struct{}{}
 				}
 			}
+
 			return util.ErrList(errList)
 		})
 	}
 	if err := errgrp.Wait(); err != nil {
-		return err
+		logrus.Errorf("[%v] error: %v", rewriteSecretsOperation, err)
+		return err // worker error from rewrites
 	}
-	log.Infof(ctx, "Cluster secrets rewritten successfully")
-	return nil
+
+	if cliErr != nil {
+		log.Infof(ctx, "[%s] Operation encountered error: %v", rewriteSecretsOperation, cliErr)
+	} else {
+		log.Infof(ctx, "[%s] Operation completed", rewriteSecretsOperation)
+	}
+
+	return cliErr
 }
 
 func (c *Cluster) RotateEncryptionKey(ctx context.Context, fullState *FullState) error {
-	//generate new key
+	// generate new key
 	newKey, err := generateEncryptionKey()
 	if err != nil {
 		return err
 	}
+
 	oldKey, err := c.extractActiveKey(c.EncryptionConfig.EncryptionProviderFile)
 	if err != nil {
 		return err
 	}
-	// reverse the keys order in the file, making newKey the Active Key
-	initialKeyList := []*encryptionKey{ // order is critical here!
-		newKey,
-		oldKey,
-	}
-	initialProviderConfig, err := providerFileFromKeyList(keyList{KeyList: initialKeyList})
+
+	logrus.Debug("adding new encryption key, provider config: [newKey, oldKey]")
+
+	// Ensure encryption is done with newKey
+	err = c.updateEncryptionProvider(ctx, []*encryptionKey{newKey, oldKey}, fullState)
 	if err != nil {
 		return err
 	}
-	c.EncryptionConfig.EncryptionProviderFile = initialProviderConfig
-	if err := c.DeployEncryptionProviderFile(ctx); err != nil {
-		return err
-	}
-	// commit to state as soon as possible
-	logrus.Debugf("[%s] Updating cluster state", services.ControlRole)
-	if err := c.UpdateClusterCurrentState(ctx, fullState); err != nil {
-		return err
-	}
-	if err := services.RestartKubeAPIWithHealthcheck(ctx, c.ControlPlaneHosts, c.LocalConnDialerFactory, c.Certificates); err != nil {
-		return err
-	}
-	// rewrite secrets
+
+	// rewrite secrets via updates to secrets
 	if err := c.RewriteSecrets(ctx); err != nil {
+		// in the case of an error during rewrite, the cluster will need to be restored, so redeploy the initial encryption provider config
+		_ = c.updateEncryptionProvider(ctx, []*encryptionKey{oldKey}, fullState)
```

Comment:
do we not want the error info here even though there was a secret rewrite error and we need to restore?

Reply:
It's not clear what should be done in the case of an error here. The way I was thinking about it was to pass the actual error from the rewrite back to the caller since that will indicate why the operation failed.

Reply:
Addressed via retries and wrapping the outer error in the case retries fail